File: backend/src/services/message.service.ts
// Contact inbox plus internal messaging permissions between admin, teacher and student roles.
import { Course, Enrollment, Message, User } from '../models/index.js';
import { MarkMessageReadInput, CreateInternalMessageInput, CreateMessageInput } from '../validations/message.validation.js';
import { HTTP_STATUS } from '../constants/http-status.js';
import { ApiError } from '../utils/api-error.js';
import { AuthUser } from '../types/auth.js';
import { emitInternalMessageReceived } from './notification-events.service.js';

function mapMessage(message: Awaited<ReturnType<typeof Message.findOne>> extends infer T ? T : never) {
  const resolved = message as NonNullable<typeof message>;
  const assignedTo = resolved.assignedTo as unknown;
  const recipientUser = resolved.recipientUser as unknown;
  const senderUser = resolved.senderUser as unknown;

  return {
    id: resolved._id.toString(),
    name: resolved.name,
    email: resolved.email,
    phone: resolved.phone,
    subject: resolved.subject,
    content: resolved.content,
    summary: resolved.summary,
    status: resolved.status,
    messageType: resolved.messageType,
    recipientRole: resolved.recipientRole,
    readAt: resolved.readAt,
    repliedAt: resolved.repliedAt,
    assignedTo:
      assignedTo && typeof assignedTo === 'object' && '_id' in (assignedTo as Record<string, unknown>)
        ? {
            id: String((assignedTo as { _id: unknown })._id),
            fullName:
              'fullName' in (assignedTo as Record<string, unknown>)
                ? String((assignedTo as { fullName?: unknown }).fullName ?? '')
                : undefined,
            email:
              'email' in (assignedTo as Record<string, unknown>)
                ? String((assignedTo as { email?: unknown }).email ?? '')
                : undefined,
          }
        : assignedTo
          ? { id: String(assignedTo) }
          : undefined,
    recipientUser:
      recipientUser && typeof recipientUser === 'object' && '_id' in (recipientUser as Record<string, unknown>)
        ? {
            id: String((recipientUser as { _id: unknown })._id),
            fullName:
              'fullName' in (recipientUser as Record<string, unknown>)
                ? String((recipientUser as { fullName?: unknown }).fullName ?? '')
                : undefined,
            email:
              'email' in (recipientUser as Record<string, unknown>)
                ? String((recipientUser as { email?: unknown }).email ?? '')
                : undefined,
            role:
              'role' in (recipientUser as Record<string, unknown>)
                ? String((recipientUser as { role?: unknown }).role ?? '')
                : undefined,
          }
        : recipientUser
          ? { id: String(recipientUser) }
          : undefined,
    senderUser:
      senderUser && typeof senderUser === 'object' && '_id' in (senderUser as Record<string, unknown>)
        ? {
            id: String((senderUser as { _id: unknown })._id),
            fullName:
              'fullName' in (senderUser as Record<string, unknown>)
                ? String((senderUser as { fullName?: unknown }).fullName ?? '')
                : undefined,
            email:
              'email' in (senderUser as Record<string, unknown>)
                ? String((senderUser as { email?: unknown }).email ?? '')
                : undefined,
            role:
              'role' in (senderUser as Record<string, unknown>)
                ? String((senderUser as { role?: unknown }).role ?? '')
                : undefined,
          }
        : senderUser
          ? { id: String(senderUser) }
          : undefined,
    createdAt: (resolved as unknown as { createdAt?: Date }).createdAt,
  };
}

async function findMessageByIdOrThrow(messageId: string) {
  const message = await Message.findById(messageId)
    .populate('assignedTo', 'fullName email role')
    .populate('recipientUser', 'fullName email role')
    .populate('senderUser', 'fullName email role');

  if (!message) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'Message not found');
  }

  return message;
}

function buildMessageSummary(content: string) {
  const normalized = content.replace(/\s+/g, ' ').trim();
  return normalized.length > 140 ? `${normalized.slice(0, 137)}...` : normalized;
}

export async function listMessages(user: AuthUser) {
  const query =
    user.role === 'admin'
      ? {}
      : {
          $or: [{ recipientUser: user.userId }, { assignedTo: user.userId }],
        };

  const messages = await Message.find(query)
    .populate('assignedTo', 'fullName email role')
    .populate('recipientUser', 'fullName email role')
    .populate('senderUser', 'fullName email role')
    .sort({ createdAt: -1 });
  return messages.map((message) => mapMessage(message));
}

export async function createMessage(input: CreateMessageInput) {
  const message = await Message.create({
    ...input,
    summary: buildMessageSummary(input.content),
    messageType: 'contact',
    status: 'unread',
  });

  const created = await findMessageByIdOrThrow(message._id.toString());
  return mapMessage(created);
}

export async function createInternalMessage(input: CreateInternalMessageInput, user: AuthUser) {
  const recipientUser = await User.findById(input.recipientUserId);

  if (!recipientUser) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'Recipient user not found');
  }

  if (user.role === 'teacher' && recipientUser.role === 'teacher') {
    throw new ApiError(HTTP_STATUS.FORBIDDEN, 'Teachers can only message admins or their students');
  }

  if (user.role === 'student' && recipientUser.role === 'student') {
    throw new ApiError(HTTP_STATUS.FORBIDDEN, 'Students can only message admins or teachers of their enrolled courses');
  }

  if (user.role === 'teacher' && recipientUser.role === 'student') {
    // Teachers can only contact students who are actually enrolled in one of their courses.
    const ownedCourseIds = await Course.find({ owner: user.userId }).distinct('_id');
    const hasSharedCourse = await Enrollment.exists({
      student: recipientUser._id,
      course: {
        $in: ownedCourseIds,
      },
      status: { $in: ['active', 'completed'] },
    });

    if (!hasSharedCourse) {
      throw new ApiError(HTTP_STATUS.FORBIDDEN, 'Teacher can only message students enrolled in their courses');
    }
  }

  if (user.role === 'student' && recipientUser.role === 'teacher') {
    // Students can only contact teachers tied to courses they currently own through enrollment.
    const enrolledCourseIds = await Enrollment.find({
      student: user.userId,
      status: { $in: ['active', 'completed'] },
    }).distinct('course');
    const teacherOwnsEnrolledCourse = await Course.exists({
      _id: { $in: enrolledCourseIds },
      owner: recipientUser._id,
    });

    if (!teacherOwnsEnrolledCourse) {
      throw new ApiError(HTTP_STATUS.FORBIDDEN, 'Student can only message teachers who own enrolled courses');
    }
  }

  const message = await Message.create({
    name: recipientUser.fullName,
    email: recipientUser.email,
    phone: recipientUser.phone,
    subject: input.subject,
    content: input.content,
    summary: buildMessageSummary(input.content),
    status: 'unread',
    messageType: 'internal',
    recipientRole: recipientUser.role,
    recipientUser: recipientUser._id,
    senderUser: user.userId,
    assignedTo: user.userId,
  });

  const created = await findMessageByIdOrThrow(message._id.toString());
  emitInternalMessageReceived({
    sender: user,
    recipientUserId: recipientUser._id.toString(),
    recipientRole: recipientUser.role,
    recipientName: recipientUser.fullName,
    subject: input.subject,
  });
  return mapMessage(created);
}

export async function listAvailableMessageRecipients(user: AuthUser) {
  if (user.role === 'admin') {
    const users = await User.find({ role: { $in: ['teacher', 'student'] }, isActive: true })
      .select('fullName email role')
      .sort({ role: 1, fullName: 1 });

    return users.map((item) => ({
      id: item._id.toString(),
      fullName: item.fullName,
      email: item.email,
      role: item.role,
    }));
  }

  if (user.role === 'teacher') {
    const [admins, studentIds] = await Promise.all([
      User.find({ role: 'admin', isActive: true }).select('fullName email role').sort({ fullName: 1 }),
      Enrollment.find({ status: { $in: ['active', 'completed'] } })
        .where('course')
        .in(await Course.find({ owner: user.userId }).distinct('_id'))
        .distinct('student'),
    ]);

    const students = studentIds.length
      ? await User.find({ _id: { $in: studentIds }, role: 'student', isActive: true })
          .select('fullName email role')
          .sort({ fullName: 1 })
      : [];

    return [...admins, ...students].map((item) => ({
      id: item._id.toString(),
      fullName: item.fullName,
      email: item.email,
      role: item.role,
    }));
  }

  if (user.role === 'student') {
    const [admins, teacherIds] = await Promise.all([
      User.find({ role: 'admin', isActive: true }).select('fullName email role').sort({ fullName: 1 }),
      Course.find({
        _id: {
          $in: await Enrollment.find({
            student: user.userId,
            status: { $in: ['active', 'completed'] },
          }).distinct('course'),
        },
      }).distinct('owner'),
    ]);

    const teachers = teacherIds.length
      ? await User.find({ _id: { $in: teacherIds }, role: 'teacher', isActive: true })
          .select('fullName email role')
          .sort({ fullName: 1 })
      : [];

    return [...admins, ...teachers].map((item) => ({
      id: item._id.toString(),
      fullName: item.fullName,
      email: item.email,
      role: item.role,
    }));
  }

  return [];
}

export async function markMessageRead(messageId: string, input: MarkMessageReadInput, user: AuthUser) {
  const message = await findMessageByIdOrThrow(messageId);
  const nextStatus = input.status ?? 'read';

  message.status = nextStatus;
  message.assignedTo = user.userId as never;
  message.readAt = nextStatus === 'read' || nextStatus === 'replied' ? new Date() : message.readAt;
  message.repliedAt = nextStatus === 'replied' ? new Date() : message.repliedAt;

  await message.save();
  const updated = await findMessageByIdOrThrow(messageId);
  return mapMessage(updated);
}

File: backend-spring/src/main/java/com/ivyts/backend/service/MessageService.java
package com.ivyts.backend.service;

import com.ivyts.backend.common.exception.ApiException;
import com.ivyts.backend.domain.course.Course;
import com.ivyts.backend.domain.enrollment.Enrollment;
import com.ivyts.backend.domain.enrollment.EnrollmentStatus;
import com.ivyts.backend.domain.message.Message;
import com.ivyts.backend.domain.message.MessageStatus;
import com.ivyts.backend.domain.message.MessageType;
import com.ivyts.backend.domain.user.User;
import com.ivyts.backend.domain.user.UserRole;
import com.ivyts.backend.notification.NotificationEventsService;
import com.ivyts.backend.security.AuthUser;
import com.ivyts.backend.service.coursestore.CourseStore;
import com.ivyts.backend.service.enrollmentstore.EnrollmentStore;
import com.ivyts.backend.service.messagestore.MessageStore;
import com.ivyts.backend.service.userstore.UserStore;
import com.ivyts.backend.web.message.MessageMapper;
import com.ivyts.backend.web.message.dto.CreateContactMessageRequest;
import com.ivyts.backend.web.message.dto.CreateInternalMessageRequest;
import com.ivyts.backend.web.message.dto.MarkMessageRequest;
import java.time.Instant;
import java.util.Comparator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

@Service
public class MessageService {

    private final MessageStore messageStore;
    private final UserStore userStore;
    private final CourseStore courseStore;
    private final EnrollmentStore enrollmentStore;
    private final MessageMapper mapper;
    private final NotificationEventsService notificationEventsService;

    public MessageService(
        MessageStore messageStore,
        UserStore userStore,
        CourseStore courseStore,
        EnrollmentStore enrollmentStore,
        MessageMapper mapper,
        NotificationEventsService notificationEventsService
    ) {
        this.messageStore = messageStore;
        this.userStore = userStore;
        this.courseStore = courseStore;
        this.enrollmentStore = enrollmentStore;
        this.mapper = mapper;
        this.notificationEventsService = notificationEventsService;
    }

    public List<Map<String, Object>> listMessages(AuthUser authUser) {
        return messageStore.findAll().stream()
            .filter(message -> canViewMessage(authUser, message))
            .sorted(Comparator.comparing(Message::getCreatedAt, Comparator.nullsLast(Comparator.reverseOrder())))
            .map(this::toMessageView)
            .toList();
    }

    public Map<String, Object> createContactMessage(CreateContactMessageRequest request) {
        Message message = new Message();
        message.setName(request.name().trim());
        message.setEmail(request.email().trim().toLowerCase(Locale.ROOT));
        message.setPhone(blankToNull(request.phone()));
        message.setSubject(request.subject().trim());
        message.setContent(request.content().trim());
        message.setSummary(buildSummary(request.content()));
        message.setMessageType(MessageType.CONTACT);
        message.setStatus(MessageStatus.UNREAD);
        return toMessageView(messageStore.save(message));
    }

    public Map<String, Object> createInternalMessage(CreateInternalMessageRequest request, AuthUser authUser) {
        User sender = findUserOrThrow(authUser.userId());
        User recipient = findUserOrThrow(request.recipientUserId());
        validateInternalMessagingPermission(sender, recipient);

        Message message = new Message();
        message.setName(recipient.getFullName());
        message.setEmail(recipient.getEmail());
        message.setPhone(recipient.getPhone());
        message.setSubject(request.subject().trim());
        message.setContent(request.content().trim());
        message.setSummary(buildSummary(request.content()));
        message.setStatus(MessageStatus.UNREAD);
        message.setMessageType(MessageType.INTERNAL);
        message.setRecipientRole(recipient.getRole());
        message.setRecipientUser(recipient.getId());
        message.setSenderUser(sender.getId());
        message.setAssignedTo(sender.getId());
        Message savedMessage = messageStore.save(message);
        notificationEventsService.emitInternalMessageReceived(
            authUser,
            recipient.getId(),
            recipient.getRole().name().toLowerCase(),
            recipient.getFullName(),
            savedMessage.getSubject()
        );
        return toMessageView(savedMessage);
    }

    public List<Map<String, Object>> listRecipients(AuthUser authUser) {
        if (authUser.role() == UserRole.ADMIN) {
            return userStore.findAll().stream()
                .filter(user -> user.isActive() && user.getRole() != UserRole.ADMIN)
                .sorted(Comparator.comparing((User user) -> user.getRole().name()).thenComparing(User::getFullName))
                .map(mapper::toRecipientView)
                .toList();
        }

        if (authUser.role() == UserRole.TEACHER) {
            List<User> admins = userStore.findAll().stream()
                .filter(user -> user.isActive() && user.getRole() == UserRole.ADMIN)
                .toList();
            List<String> ownedCourseIds = courseStore.findByOwner(authUser.userId()).stream().map(Course::getId).toList();
            LinkedHashSet<String> studentIds = enrollmentStore.findAll().stream()
                .filter(enrollment -> ownedCourseIds.contains(enrollment.getCourse()))
                .filter(enrollment -> enrollment.getStatus() == EnrollmentStatus.ACTIVE || enrollment.getStatus() == EnrollmentStatus.COMPLETED)
                .map(Enrollment::getStudent)
                .collect(java.util.stream.Collectors.toCollection(LinkedHashSet::new));
            List<User> students = studentIds.stream().map(this::findUserOrThrow).filter(User::isActive).toList();
            return java.util.stream.Stream.concat(admins.stream(), students.stream())
                .sorted(Comparator.comparing(User::getFullName))
                .map(mapper::toRecipientView)
                .toList();
        }

        if (authUser.role() == UserRole.STUDENT) {
            List<User> admins = userStore.findAll().stream()
                .filter(user -> user.isActive() && user.getRole() == UserRole.ADMIN)
                .toList();
            List<String> enrolledCourseIds = enrollmentStore.findByStudentAndStatusInOrderByCreatedAtDesc(
                authUser.userId(),
                List.of(EnrollmentStatus.ACTIVE, EnrollmentStatus.COMPLETED)
            ).stream().map(Enrollment::getCourse).distinct().toList();
            LinkedHashSet<String> teacherIds = courseStore.findAllByIds(enrolledCourseIds).stream()
                .map(Course::getOwner)
                .collect(java.util.stream.Collectors.toCollection(LinkedHashSet::new));
            List<User> teachers = teacherIds.stream().map(this::findUserOrThrow).filter(User::isActive).toList();
            return java.util.stream.Stream.concat(admins.stream(), teachers.stream())
                .sorted(Comparator.comparing(User::getFullName))
                .map(mapper::toRecipientView)
                .toList();
        }

        return List.of();
    }

    public Map<String, Object> markMessage(String messageId, MarkMessageRequest request, AuthUser authUser) {
        Message message = findMessageOrThrow(messageId);
        if (!canViewMessage(authUser, message)) {
            throw new ApiException(HttpStatus.NOT_FOUND, "Message not found");
        }

        MessageStatus nextStatus = parseStatus(request.status());
        message.setStatus(nextStatus);
        message.setAssignedTo(authUser.userId());
        if (nextStatus == MessageStatus.READ || nextStatus == MessageStatus.REPLIED) {
            message.setReadAt(Instant.now());
        }
        if (nextStatus == MessageStatus.REPLIED) {
            message.setRepliedAt(Instant.now());
        }
        return toMessageView(messageStore.save(message));
    }

    private boolean canViewMessage(AuthUser authUser, Message message) {
        if (authUser.role() == UserRole.ADMIN) {
            return true;
        }
        return authUser.userId().equals(message.getRecipientUser()) || authUser.userId().equals(message.getAssignedTo());
    }

    private void validateInternalMessagingPermission(User sender, User recipient) {
        if (sender.getRole() == UserRole.ADMIN) {
            return;
        }

        if (sender.getRole() == UserRole.TEACHER) {
            if (recipient.getRole() == UserRole.TEACHER) {
                throw new ApiException(HttpStatus.FORBIDDEN, "Teachers can only message admins or their students");
            }
            if (recipient.getRole() == UserRole.STUDENT) {
                List<String> ownedCourseIds = courseStore.findByOwner(sender.getId()).stream().map(Course::getId).toList();
                boolean hasSharedCourse = enrollmentStore.findAll().stream()
                    .anyMatch(enrollment -> enrollment.getStudent().equals(recipient.getId())
                        && ownedCourseIds.contains(enrollment.getCourse())
                        && (enrollment.getStatus() == EnrollmentStatus.ACTIVE || enrollment.getStatus() == EnrollmentStatus.COMPLETED));
                if (!hasSharedCourse) {
                    throw new ApiException(HttpStatus.FORBIDDEN, "Teacher can only message students enrolled in their courses");
                }
            }
            return;
        }

        if (sender.getRole() == UserRole.STUDENT) {
            if (recipient.getRole() == UserRole.STUDENT) {
                throw new ApiException(HttpStatus.FORBIDDEN, "Students can only message admins or teachers of enrolled courses");
            }
            if (recipient.getRole() == UserRole.TEACHER) {
                List<String> enrolledCourseIds = enrollmentStore.findByStudentAndStatusInOrderByCreatedAtDesc(
                    sender.getId(),
                    List.of(EnrollmentStatus.ACTIVE, EnrollmentStatus.COMPLETED)
                ).stream().map(Enrollment::getCourse).toList();
                boolean teacherOwnsEnrolledCourse = courseStore.findAllByIds(enrolledCourseIds).stream()
                    .anyMatch(course -> course.getOwner().equals(recipient.getId()));
                if (!teacherOwnsEnrolledCourse) {
                    throw new ApiException(HttpStatus.FORBIDDEN, "Student can only message teachers who own enrolled courses");
                }
            }
        }
    }

    private Message findMessageOrThrow(String messageId) {
        return messageStore.findById(messageId)
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Message not found"));
    }

    private User findUserOrThrow(String userId) {
        return userStore.findById(userId)
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "User not found"));
    }

    private Map<String, Object> toMessageView(Message message) {
        User assignedTo = message.getAssignedTo() != null ? userStore.findById(message.getAssignedTo()).orElse(null) : null;
        User recipient = message.getRecipientUser() != null ? userStore.findById(message.getRecipientUser()).orElse(null) : null;
        User sender = message.getSenderUser() != null ? userStore.findById(message.getSenderUser()).orElse(null) : null;
        return mapper.toMessageView(message, assignedTo, recipient, sender);
    }

    private String buildSummary(String content) {
        String normalized = content.replaceAll("\\s+", " ").trim();
        return normalized.length() > 140 ? normalized.substring(0, 137) + "..." : normalized;
    }

    private MessageStatus parseStatus(String value) {
        String normalized = value == null || value.isBlank() ? "read" : value.trim().replace('-', '_').toUpperCase(Locale.ROOT);
        if (!normalized.equals("READ") && !normalized.equals("REPLIED")) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "Status must be read or replied");
        }
        return MessageStatus.valueOf(normalized);
    }

    private String blankToNull(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }
}
